*** gcc-4.1.0/./gcc/config/i386/i386-protos.h	2006-09-16 15:25:39.000000000 +0200
--- gcc-4.1.0-patch-segreg/./gcc/config/i386/i386-protos.h	2006-09-16 20:03:22.000000000 +0200
*************** extern rtx maybe_get_pool_constant (rtx)
*** 209,215 ****
  extern char internal_label_prefix[16];
  extern int internal_label_prefix_len;
  
! enum ix86_address_seg { SEG_DEFAULT, SEG_FS, SEG_GS };
  struct ix86_address
  {
    rtx base, index, disp;
--- 209,215 ----
  extern char internal_label_prefix[16];
  extern int internal_label_prefix_len;
  
! enum ix86_address_seg { SEG_DEFAULT, SEG_CS, SEG_SS, SEG_DS, SEG_ES, SEG_FS, SEG_GS };
  struct ix86_address
  {
    rtx base, index, disp;
