*** gcc-4.1.0/gcc/config/i386/i386-protos.h	2005-10-19 10:13:37.000000000 +0800
--- gcc-4.1.0-patch-far/gcc/config/i386/i386-protos.h	2006-05-30 19:05:46.000000000 +0800
*************** extern rtx maybe_get_pool_constant (rtx)
*** 209,215 ****
  extern char internal_label_prefix[16];
  extern int internal_label_prefix_len;
  
! enum ix86_address_seg { SEG_DEFAULT, SEG_FS, SEG_GS };
  struct ix86_address
  {
    rtx base, index, disp;
--- 209,215 ----
  extern char internal_label_prefix[16];
  extern int internal_label_prefix_len;
  
! enum ix86_address_seg { SEG_DEFAULT, SEG_SS, SEG_DS, SEG_ES, SEG_FS, SEG_GS };
  struct ix86_address
  {
    rtx base, index, disp;
